src/net_processing.cpp

```diff
@@ -3195,6 +3210,115 @@ void PeerManagerImpl::ProcessValidTx(NodeId nodeid, const CTransactionRef& tx, c
+        if (Assume(it_result != package_result.m_tx_results.end())) {
+            const auto& tx_result = it_result->second;
+            switch (tx_result.m_result_type) {
+                case MempoolAcceptResult::ResultType::VALID:
+                {
+                    Assume(tx_result.m_replaced_transactions.has_value());
+                    std::list<CTransactionRef> empty_replacement_list;
+                    ProcessValidTx(nodeid, tx, tx_result.m_replaced_transactions.value_or(empty_replacement_list));
+                    break;
+                }
+                case MempoolAcceptResult::ResultType::INVALID:
+                case MempoolAcceptResult::ResultType::DIFFERENT_WITNESS:
+                {
+                    // Don't add to vExtraTxnForCompact, as these transactions should have already been
+                    // added there when added to the orphanage or rejected for TX_RECONSIDERABLE.
+                    // This should be updated if package submission is ever used for transactions
+                    // that haven't already been validated before.
+                    ProcessInvalidTx(nodeid, tx, tx_result.m_state, /*maybe_add_extra_compact_tx=*/false);
+                    break;
+                }
+                case MempoolAcceptResult::ResultType::MEMPOOL_ENTRY:
+                {
+                    // AlreadyHaveTx() should be catching transactions that are already in mempool.
+                    Assume(false);
+                    break;
+                }
+            }
+        }
+        package_iter++;
+        // There should be a sender for each tx, but fall back to the last one otherwise.
+        if (Assume(senders_iter != senders.rend())) senders_iter++;
+    }
+}
+
+void PeerManagerImpl::MaybeProcess1P1CPackage(const CTransactionRef& ptx, NodeId nodeid)
+{
+    AssertLockNotHeld(m_peer_mutex);
+    AssertLockHeld(g_msgproc_mutex);
+    AssertLockHeld(cs_main);
+
+    const auto& parent_txid{ptx->GetHash()};
+    const auto& parent_wtxid{ptx->GetWitnessHash()};
+
+    const auto cpfp_candidates{m_orphanage.GetChildren(ptx)};
+    // Exit early.
+    if (cpfp_candidates.empty()) {
+        LogDebug(BCLog::TXPACKAGES, "didn't evaluate package for %s (wtxid=%s), no corresponding orphan found\n",
+                 parent_txid.ToString(), parent_wtxid.ToString());
+        return;
+    }
+
+    CTransactionRef tx_orphan{nullptr};
+    NodeId orphan_sender{-1};
+    Assume(m_recent_rejects_reconsiderable.contains(parent_wtxid.ToUint256()));
+
+    // Create a 1p1c package using the first child that hasn't already been rejected. Sort
+    // the children in random order to not create a bias that attackers can use to delay
+    // package acceptance. Create a random permutation of the indices.
+    std::vector<size_t> tx_indices(cpfp_candidates.size());
+    std::iota(tx_indices.begin(), tx_indices.end(), 0);
+    Shuffle(tx_indices.begin(), tx_indices.end(), m_rng);
```

Comment:
would it make sense to match peers and (if possible) always pick a child from the orphanage that was sent to us by the same peer that sent us the parent (instead of a random one)?
That way, it wouldn't be possible that a third peer could send us multiple low-fee children that we'd store in the orphanage, in the hope that we pick one of those and reject the package.

Reply:
I think I made this point in person from the philosophical standpoint that orphanage-churning aside, we probably shouldn't allow peers to "cross-talk" when it comes to package evaluation and possible punishment.

In practice I think an adversary can just churn the orphanage, but still might be a good conceptual framework to adhere to?

Reply:
great idea, will change 👍

Reply:
For some code deduplication you could just have `GetChildrenFromPeer` which takes `std::optional<NodeId>`, returning full `std::vector<std::pair<CTransactionRef, NodeId>>` list if `std::nullopt` or filtered to the specific nodeid otherwise. You can call it filtered once, then if a suitable orphan isn't found, call it again unfiltered.

Reply:
I don't think that is much simpler. Also I think this is easier to delete if/when we don't try orphans from different peers in the future.